### ISSUE TITLE
wxGrid should disable ime input on read-only cells

### ISSUE THREAD

**Issue by marekr** [wxWidgets/wxWidgets#24558]
Currently, if you click on a wxGrid cell that is read-only and then start typing. 
The IME will open in the far right corner of your display, in my case, on the other monitor

![image](https://github.com/wxWidgets/wxWidgets/assets/1838013/928d06ce-13ae-4f5e-a63f-eb02c5b791d4)

If you click into editable cells, the IME will correctly show under that cell

![image](https://github.com/wxWidgets/wxWidgets/assets/1838013/0c62c269-90db-4866-9386-ed85d5db9810)

Doing so should be as simple as 
`::ImmAssociateContext( Hwnd, NULL );`

but it probably has to be one one of the wxGrid sub windows that make up the entire thing

**Comment by vadz** [comment]
Hmm, I must be using a wrong locale because I don't see anything like this at all. I have "Chinese (Traditional, Taiwan) Microsoft Bopomofo" in the language switcher, which one do you use?

**Comment by marekr** [comment]
> Hmm, I must be using a wrong locale because I don't see anything like this at all. I have "Chinese (Traditional, Taiwan) Microsoft Bopomofo" in the language switcher, which one do you use?

Chinese traditional, ill explore further

**Comment by vadz** [comment]
Yes, I thought I installed the same locale that you used, but somehow I don't see this window at all. Do I need to press anything special to make it come up? I've tried checking if there're some options I could change, but I don't see any and after doing "Restore default IME settings" nothing seems to have changed.

**Comment by marekr** [comment]
I can reproduce it in the grid sample.

Simply added

    grid->SetReadOnly(0, 0);

below

    grid->SetCellValue( 0, 0, "Ctrl+Home\nwill go to\nthis cell" );


When that corner cell is selected, and you start typing with 

![lCXO49Zwxn](https://github.com/wxWidgets/wxWidgets/assets/1838013/7a257441-62c8-449a-a215-96e5152134ad)

**Comment by marekr** [comment]
This fixes it for me

 include/wx/generic/private/grid.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/include/wx/generic/private/grid.h b/include/wx/generic/private/grid.h
index 9ada5661c0..e8cb2e5f41 100644
--- a/include/wx/generic/private/grid.h
+++ b/include/wx/generic/private/grid.h
@@ -392,6 +392,9 @@ public:
           m_type(type)
     {
         SetBackgroundStyle(wxBG_STYLE_PAINT);
+#ifdef __WXMSW__
+        ::ImmAssociateContext(GetHWND(), NULL);
+#endif
     }
 
 


The text input in cells still shows IME, the rest of the wxGrid won't show the IME

**Comment by vadz** [comment]
Thanks, I do see this with simplified Chinese (I still have no idea why does the IME for simplified and traditional Chinese work so differently...).

However I don't think disabling IME for the grid is the best solution because you should be able to start typing and make the cell editor appear with the character you've entered already in it — and this actually works even with the IME right now. With your patch, it doesn't work any more however.

Ideal would be to avoid showing the IME window only for the read-only cells, I'm going to see if I find a way to do it.

BTW, I suspect this problem is not MSW-specific neither.

**Comment by marekr** [comment]
> Thanks, I do see this with simplified Chinese (I still have no idea why does the IME for simplified and traditional Chinese work so differently...).
> 
> However I don't think disabling IME for the grid is the best solution because you should be able to start typing and make the cell editor appear with the character you've entered already in it — and this actually works even with the IME right now. With your patch, it doesn't work any more however.
> 
> Ideal would be to avoid showing the IME window only for the read-only cells, I'm going to see if I find a way to do it.
> 
> BTW, I suspect this problem is not MSW-specific neither.

Then it needs a two part fix .

Read-only cells need the ime disabled.

But also editable cells that don't have the editor triggered yet need the ime target set manually using another imm function because it also points the editor in space.

**Comment by vadz** [comment]
I can prevent IME window from appearing (I think this one is the "candidate" window, right?) by not passing `WM_KEYDOWN` to `DefWindowProc()`, but I'd like to find some more precise way of doing it. Unfortunately [WM_IME_NOTIFY](https://learn.microsoft.com/en-us/windows/win32/intl/wm-ime-notify) doesn't seem to allow returning something to prevent this from happening and I don't know what else to try right now.

Positioning the window can be done with [ImmSetCompositionWindow()](https://learn.microsoft.com/en-us/windows/win32/api/imm/nf-imm-immsetcompositionwindow) but I'm not sure when to call it neither.

**Comment by vadz** [comment]
I'll try to do something about it before the next release. If anybody else has any hints/better ideas about how to fix this, please let me know.

**Comment by vadz** [comment]
Sorry, I failed to fix this in reasonable amount of time, this needs more work as there are problems with positioning of the IME window too (even for normal cells). I hope to return to this later...

**Comment by vadz** [comment]
I've finally got back to this and (a little help from LLM with setting up and testing Chinese IME was very welcome) got something working, see #27026. It's still not finished because we need to do something about activatable cells (checkboxes) too, but it already fixes the problems mentioned here and testing it would be very welcome!
